sentinelhub/evalscript.py

```diff
@@ -0,0 +1,117 @@
+"""
+Module defining evalscript generation utilities
+"""
+from typing import Dict, List, Optional
+
+import numpy as np
+
+from .data_collections import DataCollection
+from .data_collections_bands import Band, Unit
+
+DTYPE_TO_SAMPLE_TYPE: Dict[type, str] = {
+    bool: "UINT8",
+    np.uint8: "UINT8",
+    np.uint16: "UINT16",
+    np.float32: "FLOAT32",
+}
+
+EVALSCRIPT_TEMPLATE = """
+    //VERSION=3
+
+    function setup() {{
+        return {{
+            input: [{{
+                bands: [{input_names}],
+                units: [{input_units}]
+            }}],
+            output: [{output_spec}]
+        }}
+    }}
+
+    function updateOutputMetadata(scenes, inputMetadata, outputMetadata) {{
+        outputMetadata.userData = {{
+            "norm_factor":  inputMetadata.normalizationFactor
+        }}
+    }}
+
+    function evaluatePixel(sample) {{
+        return {{ {return_spec} }};
+    }}
+"""
+
+
+def parse_data_collection_bands(data_collection: DataCollection, bands: List[str]) -> List[Band]:
+    """Checks that all requested bands are available and returns the band information for further processing
+
+    :param data_collection: A collection of requested satellite data.
+    :param bands: A list of band or meta band names to use in the evalscript.
+    """
+
+    requested_bands = []
+    band_info_dict = {band_info.name: band_info for band_info in data_collection.bands + data_collection.metabands}
+
+    for band_name in bands:
+        if band_name not in band_info_dict:
+            raise ValueError(
+                f"Data collection {data_collection} does not have specifications for band {band_name}.\n"
+                f"Available bands are:\n{[band.name for band in data_collection.bands]}\nand meta-bands:\n"
+                f"{[band.name for band in data_collection.metabands]}."
+            )
+        requested_bands.append(band_info_dict[band_name])
+
+    return requested_bands
+
+
+def generate_evalscript(
+    data_collection: DataCollection,
+    bands: Optional[List[str]] = None,
+    meta_bands: Optional[List[str]] = None,
+    merged_output: Optional[str] = None,
+    use_dn: bool = True,
```

Comment:
This only allows us to switch to `DN` but not other units right?
To be fair ATM it is very rare to even have two options for units, but this way the user can't get bands + `sunAzimuthAngles` because those only have degrees.

I know it's a weird edge case, but the way this is now it's not a direct replacement for `eo-learn`. Should we do a direct replacement? Or do we make something else and wait for new tasks to replace SHIT

Reply:
ah, i see from the docs that it is more of a 'force_dn_when_possible`
I do like the idea going forward, just sad that SHIT is as it is :(

Reply:
> ah, i see from the docs that it is more of a 'force_dn_when_possible`

Yes, correct, the flag means to `use DN if possible, otherwise the first default choice`. Do you think the name of the parameter is too misleading? Perhaps it could be `dn_priority` or `prioritize_dn` or something like that

> This only allows us to switch to DN but not other units right?

That seemed like a too specific case to me, so in this situation it would be up to the user to provide a custom defined evalscript.

> I know it's a weird edge case, but the way this is now it's not a direct replacement for eo-learn.

> , just sad that SHIT is as it is :(

Could you elaborate what you mean exactly? I see no issue with the above being the direct replacement for `eo-learn`. Or perhaps that was still the old comment.

Reply:
it's not a direct replacement, it would feature some API changes
but i reckon that would be small enough to be worth it